File: articles/language-development-and-the-social-formation-of-speech/r/model_language_development.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
script_path <- normalizePath(sub("--file=", "", args[grep("--file=", args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
data_path <- file.path(root, "data", "language_development_panel.csv")
outputs_dir <- file.path(root, "outputs")
if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)

panel <- read.csv(data_path)

model <- lmer(
  language_score ~ time + I(time^2) + current_interaction +
    current_reading + current_joint_attention + current_turn_taking +
    hearing_support + multilingual_exposure + current_stress +
    chronic_stress + language_ecology_support + book_access +
    early_education_quality + home_language_recognition +
    multilingual_exposure:home_language_recognition +
    language_support_context +
    (1 + time | context_id/child_id),
  data = panel
)

capture.output(summary(model), file = file.path(outputs_dir, "r_language_development_model_summary.txt"))

trajectory <- panel |>
  group_by(time, chronic_stress) |>
  summarize(
    average_language = mean(language_score),
    average_interaction = mean(current_interaction),
    average_reading = mean(current_reading),
    average_joint_attention = mean(current_joint_attention),
    average_turn_taking = mean(current_turn_taking),
    average_stress = mean(current_stress),
    average_language_support = mean(language_support_context),
    standard_error = sd(language_score) / sqrt(n()),
    .groups = "drop"
  ) |>
  mutate(
    stress_group = ifelse(chronic_stress == 1, "Higher chronic stress", "Lower chronic stress")
  )

write.csv(trajectory, file.path(outputs_dir, "r_stress_language_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_stress_language_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_language, linetype = stress_group)) +
    geom_line(linewidth = 1) +
    labs(
      title = "Synthetic Language Development Under Support and Stress",
      x = "Time",
      y = "Average language score",
      linetype = "Group"
    ) +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R language development outputs.")

File: articles/language-development-and-the-social-formation-of-speech/python/model_language_development.py
#!/usr/bin/env python3
from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "language_development_panel.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

def main() -> None:
    panel = pd.read_csv(DATA).sort_values(["child_id", "time"])
    panel["lag_score"] = panel.groupby("child_id")["language_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    model = smf.ols(
        """
        language_score ~ lag_score + time + I(time ** 2) +
        current_interaction + current_reading + current_joint_attention +
        current_turn_taking + hearing_support + multilingual_exposure +
        current_stress + chronic_stress + language_ecology_support +
        book_access + early_education_quality + home_language_recognition +
        multilingual_exposure:home_language_recognition +
        language_support_context
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    profile_model = smf.ols(
        """
        language_score ~ time + I(time ** 2) + chronic_stress +
        current_stress + current_interaction + current_reading +
        language_support_context + C(language_profile)
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUT / "python_language_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC LANGUAGE DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(model.summary().as_text())
        f.write("\n\nLANGUAGE PROFILE MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(profile_model.summary().as_text())

    trajectory = panel.groupby(["time", "chronic_stress"], as_index=False).agg(
        average_language=("language_score", "mean"),
        average_interaction=("current_interaction", "mean"),
        average_reading=("current_reading", "mean"),
        average_joint_attention=("current_joint_attention", "mean"),
        average_turn_taking=("current_turn_taking", "mean"),
        average_stress=("current_stress", "mean"),
        average_language_support=("language_support_context", "mean"),
        standard_error=("language_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["stress_group"] = trajectory["chronic_stress"].map({
        0: "Lower chronic stress",
        1: "Higher chronic stress",
    })
    trajectory.to_csv(OUT / "python_stress_language_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for label, group in trajectory.groupby("stress_group"):
        plt.plot(group["time"], group["average_language"], marker="o", label=label)
    plt.xlabel("Time")
    plt.ylabel("Average language score")
    plt.title("Synthetic Language Development Under Support and Stress")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUT / "python_stress_language_trajectory.png", dpi=160)
    plt.close()

    context = panel.groupby("time", as_index=False).agg(
        average_interaction=("current_interaction", "mean"),
        average_reading=("current_reading", "mean"),
        average_joint_attention=("current_joint_attention", "mean"),
        average_turn_taking=("current_turn_taking", "mean"),
        average_stress=("current_stress", "mean"),
        average_language_support=("language_support_context", "mean"),
        average_language=("language_score", "mean"),
    )
    context.to_csv(OUT / "python_language_context_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for col, label in [
        ("average_interaction", "responsive interaction"),
        ("average_reading", "shared reading"),
        ("average_joint_attention", "joint attention"),
        ("average_turn_taking", "turn-taking"),
        ("average_stress", "stress"),
        ("average_language_support", "language support"),
    ]:
        plt.plot(context["time"], context[col], marker="o", label=label)
    plt.xlabel("Time")
    plt.ylabel("Average index")
    plt.title("Synthetic Language Context Across Time")
    plt.legend(fontsize=8)
    plt.tight_layout()
    plt.savefig(OUT / "python_language_context_trajectory.png", dpi=160)
    plt.close()

    panel.groupby("language_profile", as_index=False).agg(
        children=("child_id", "nunique"),
        average_language=("language_score", "mean"),
        average_language_support_context=("language_support_context", "mean"),
        average_interaction=("current_interaction", "mean"),
        average_reading=("current_reading", "mean"),
        average_joint_attention=("current_joint_attention", "mean"),
        average_turn_taking=("current_turn_taking", "mean"),
        average_stress=("current_stress", "mean"),
        chronic_stress_rate=("chronic_stress", "mean"),
        multilingual_exposure_rate=("multilingual_exposure", "mean"),
    ).to_csv(OUT / "python_language_development_profiles.csv", index=False)

    panel.groupby("context_id", as_index=False).agg(
        language_ecology_support=("language_ecology_support", "mean"),
        book_access=("book_access", "mean"),
        early_education_quality=("early_education_quality", "mean"),
        home_language_recognition=("home_language_recognition", "mean"),
        average_language=("language_score", "mean"),
        average_stress=("current_stress", "mean"),
        average_language_support=("language_support_context", "mean"),
    ).to_csv(OUT / "python_language_context_summary.csv", index=False)

    print("Wrote Python language model outputs.")

if __name__ == "__main__":
    main()
